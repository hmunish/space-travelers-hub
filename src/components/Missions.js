import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchMissions,
  joinMission,
  cancelMission,
} from '../redux/missions/missionSlice';
import './mission.css';

const Missions = () => {
  const dispatch = useDispatch();
  const {
    missions: MissionsVar,
    isLoading,
    isError,
  } = useSelector((store) => store.missionsList);
  const storedMissions = JSON.parse(localStorage.getItem('missions')) || [];
  const missions = storedMissions.length > 0 ? storedMissions : MissionsVar;

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);
  if (isLoading) {
    return (
      <div className="loading">
        <h1>Loading...</h1>
      </div>
    );
  }
  if (isError) {
    return (
      <div className="error">
        <h1>Something went wrong...</h1>
      </div>
    );
  }

  return (
    <div className="container">
      <div className="table-div">
        <table className="table">
          <thead className="table-row">
            <tr className="head">
              <th className="name-colum">Mission</th>
              <th className="description-colum">Description</th>
              <th>Status</th>
              <th>&nbsp;</th>
            </tr>
          </thead>
          <tbody className="t-body">
            {missions.map((mission) => (
              <tr key={mission.id} className="row">
                <td className="name-colum">{mission.name}</td>
                <td className="description-colum">{mission.description}</td>
                <td>
                  {!mission.joined && (
                    <span className="not-member">NOT A MEMBER</span>
                  )}
                  {mission.joined && (
                    <span className="member">Active Member</span>
                  )}
                </td>
                <td className="text-center">
                  {mission.joined && (
                    <button
                      type="button"
                      className="btn-cancel"
                      onClick={() => dispatch(cancelMission(mission.id))}
                    >
                      Leave Mission
                    </button>
                  )}
                  {!mission.joined && (
                    <button
                      type="button"
                      className="btn-join"
                      onClick={() => dispatch(joinMission(mission.id))}
                    >
                      Join Mission
                    </button>
                  )}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};
export default Missions;
