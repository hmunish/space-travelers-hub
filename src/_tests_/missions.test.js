import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Missions from '../components/Missions';

const mockStore = configureStore([]);

describe('test Missions component', () => {
  it('should render correct snapshot', () => {
    const mockMissions = [
      {
        id: 1,
        name: 'Mission 1',
        description: 'This is mission 1',
        joined: true,
      },
      {
        id: 2,
        name: 'Mission 2',
        description: 'This is mission 2',
        joined: false,
      },
    ];

    const store = mockStore({
      missionsList: {
        missions: mockMissions,
        isLoading: false,
        isError: false,
      },
    });

    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
