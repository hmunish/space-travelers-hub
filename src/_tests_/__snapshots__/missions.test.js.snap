// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Missions component should render correct snapshot 1`] = `
<div
  className="container"
>
  <div
    className="table-div"
  >
    <table
      className="table"
    >
      <thead
        className="table-row"
      >
        <tr
          className="head"
        >
          <th
            className="name-colum"
          >
            Mission
          </th>
          <th
            className="description-colum"
          >
            Description
          </th>
          <th
            className="status"
          >
            Status
          </th>
          <th
            className="space"
          >
            Â 
          </th>
        </tr>
      </thead>
      <tbody
        className="t-body"
      >
        <tr
          className="row"
        >
          <td
            className="name-colum"
          >
            Mission 1
          </td>
          <td
            className="description-colum"
          >
            This is mission 1
          </td>
          <td
            className="status"
          >
            <span
              className="member"
            >
              Active Member
            </span>
          </td>
          <td
            className="space"
          >
            <button
              className="btn-cancel"
              onClick={[Function]}
              type="button"
            >
              Leave Mission
            </button>
          </td>
        </tr>
        <tr
          className="row"
        >
          <td
            className="name-colum"
          >
            Mission 2
          </td>
          <td
            className="description-colum"
          >
            This is mission 2
          </td>
          <td
            className="status"
          >
            <span
              className="not-member"
            >
              NOT A MEMBER
            </span>
          </td>
          <td
            className="space"
          >
            <button
              className="btn-join"
              onClick={[Function]}
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
